### PR TITLE
handle 4xx errors specifically in calls from service workers

### PR DESCRIPTION
**Pull request by johnduffell**
## Why are you doing this?

A while ago I was investigatng a step function that timed out after 24 hours.  It turned out there was a 4xx error occurring in one of the backends, i.e. client error.  I think something was missing (I forget the exact details).
It should have aborted straight away, but the step functions were turning any error when doing an http request as a Helper error, which was turning into a RetryUnlimited (i.e. 24 hours).

This PR separates out client errors (4xx) from general explosions, so that 4xx are always aborted and any other failure is retried.

Tested locally via it testing and remotely in CODE (GW 6 for 6 and contribution)

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -7,7 +7,8 @@ skip in publish := true
 lazy val integrationTestSettings: Seq[Def.Setting[_]] = Defaults.itSettings ++ Seq(
   scalaSource in IntegrationTest := baseDirectory.value / "src" / "test" / "scala",
   javaSource in IntegrationTest := baseDirectory.value / "src" / "test" / "java",
-  testOptions in Test := Seq(Tests.Argument(TestFrameworks.ScalaTest, "-l", "com.gu.test.tags.annotations.IntegrationTest"))
+  testOptions in Test := Seq(Tests.Argument(TestFrameworks.ScalaTest, "-l", "com.gu.test.tags.annotations.IntegrationTest")),
+  parallelExecution in IntegrationTest := false
```
Comment (johnduffell) [inline build.sbt]: I did this because the it tests were having timeouts in futures, however I think giving sbt 2GB instead of 1.5GB sped things up.  I'll play with removing that again tomorrow.

File: support-services/src/main/scala/com/gu/helpers/WebServiceHelper.scala [hunk 2]
```diff
@@ -11,9 +11,23 @@ import scala.collection.immutable.Map.empty
...
+import scala.util.{Failure, Success, Try}
 
-case class WebServiceHelperError[T: ClassTag](responseCode: Int, responseBody: String) extends Throwable {
-  override def getMessage: String = s"${classTag[T]} - $responseCode: $responseBody"
+sealed abstract class WebServiceHelperErrorBase(responseCode: String, responseBody: String) extends Throwable
+
+case class WebServiceHelperError[T: ClassTag](
+  responseCode: String,
+  responseBody: String,
+  message: String
```
Comment (johnduffell) [inline support-services/src/main/scala/com/gu/helpers/WebServiceHelper.scala]: I added a message so we can be more clear what went wrong in the logs.

File: support-services/src/main/scala/com/gu/helpers/WebServiceHelper.scala [hunk 3]
```diff
@@ -11,9 +11,23 @@ import scala.collection.immutable.Map.empty
...
+) extends WebServiceHelperErrorBase(responseCode, responseBody) {
+  override def getMessage: String = s"${classTag[T]} - $message: $responseCode: $responseBody"
+}
+
+case class WebServiceClientError(
```
Comment (johnduffell) [inline support-services/src/main/scala/com/gu/helpers/WebServiceHelper.scala]: this covers all 4xx errors

File: support-services/src/main/scala/com/gu/helpers/WebServiceHelper.scala [hunk 4]
```diff
@@ -11,9 +11,23 @@ import scala.collection.immutable.Map.empty
...
 import scala.concurrent.Future
 import scala.reflect.{ClassTag, classTag}
+import scala.util.{Failure, Success, Try}
 
-case class WebServiceHelperError[T: ClassTag](responseCode: Int, responseBody: String) extends Throwable {
-  override def getMessage: String = s"${classTag[T]} - $responseCode: $responseBody"
+sealed abstract class WebServiceHelperErrorBase(responseCode: String, responseBody: String) extends Throwable
+
+case class WebServiceHelperError[T: ClassTag](
+  responseCode: String,
```
Comment (johnduffell) [inline support-services/src/main/scala/com/gu/helpers/WebServiceHelper.scala]: in reality is isn't really a String or an Int, but I changed to string so I can index into it easily

File: support-services/src/main/scala/com/gu/helpers/WebServiceHelper.scala [hunk 5]
```diff
@@ -51,19 +65,32 @@ trait WebServiceHelper[Error <: Throwable] {
    */
   private def request[A](rb: Request.Builder)(implicit decoder: Decoder[A], errorDecoder: Decoder[Error], ctag: ClassTag[A]): Future[A] = {
     val req = wsPreExecute(rb).build()
-    SafeLogger.debug(s"Issuing request ${req.method} ${req.url}")
+    SafeLogger.info(s"Issuing request ${req.method} ${req.url}")
     for {
       response <- httpClient(req)
-    } yield {
-      val responseBody = response.body.string()
```
Comment (johnduffell) [inline support-services/src/main/scala/com/gu/helpers/WebServiceHelper.scala]: body and string are both nullable according to okhttp3 docs

File: support-services/src/main/scala/com/gu/helpers/WebServiceHelper.scala [hunk 6]
```diff
@@ -51,19 +65,32 @@ trait WebServiceHelper[Error <: Throwable] {
...
+        val code = response.code().toString
+        val failableDecodedResponse = for {
+          body <- Option(response.body).toRight(WebServiceHelperError(code, "", "no body")).toTry
+          contentType <- Option(body.contentType()).toRight(WebServiceHelperError(code, "", "no content type")).toTry
+          bodyString <- Try(body.string())
+          _ = SafeLogger.info(s"response $code body: ${bodyString.length} bytes")
+          _ <- if (contentType.`type`() == "application" && contentType.subtype() == "json") Success(())
+          else Failure(WebServiceHelperError(code, bodyString, s"wrong content type"))
+          result <- code.splitAt(1) match {
+            case ("2", _) => decode[A](bodyString).left.map { err =>
```
Comment (jfsoul) [inline support-services/src/main/scala/com/gu/helpers/WebServiceHelper.scala]: Could have left `code` as an `int` and then we don't have to split a string and match on that.  There's also an `isSuccessful` method by the looks of it, but no equivalent for client errors.

File: support-workers/src/main/scala/com/gu/support/workers/exceptions/RetryImplicits.scala [hunk 7]
```diff
@@ -15,8 +15,8 @@ object RetryImplicits {
       //Timeouts/connection issues and 500s
       case e @ (_: SocketTimeoutException | _: SocketException | _: WebServiceHelperError[_]) => new RetryUnlimited(message = e.getMessage, cause = throwable)
 
-      //Invalid Json
-      case e @ (_: ParsingFailure | _: MatchError) => new RetryNone(message = e.getMessage, cause = throwable)
+      //Invalid Json or http client error (4xx)
+      case e @ (_: ParsingFailure | _: MatchError | _: WebServiceClientError) => new RetryNone(message = e.getMessage, cause = throwable)
```
Comment (jfsoul) [inline support-workers/src/main/scala/com/gu/support/workers/exceptions/RetryImplicits.scala]: We're sure that there aren't any retryable 4XX errors?

Comment (johnduffell) [reply support-workers/src/main/scala/com/gu/support/workers/exceptions/RetryImplicits.scala]: well nothing is for sure, especially with zuora.  Having said that I have a feeling zuora errors are returned as 2xx with success=false or Success=false in the body, so I will investigate (and add tests)

File: support-services/src/main/scala/com/gu/helpers/WebServiceHelper.scala (hunk 6 above)
Comment (johnduffell) [reply support-services/src/main/scala/com/gu/helpers/WebServiceHelper.scala]: yeah isSuccessful seems to be the only one in okhttp, unfortunately no nice ADT of response codes.
I agree I could have left as a int, but I'd still have to do `code % 100` and check it against 2 and 4, but I don't entirely feel like arithmetic makes sense on a response code.  More concretely it opens us up to off by one errors.  In a string each character is opaque when splitting so if it works in one case it should work in them all.
that's my reasoning but happy to use mod instead if you feel strongly!

File: support-services/src/main/scala/com/gu/helpers/WebServiceHelper.scala [hunk 8]
```diff
@@ -51,19 +65,32 @@ trait WebServiceHelper[Error <: Throwable] {
...
       response <- httpClient(req)
-    } yield {
-      val responseBody = response.body.string()
-      SafeLogger.debug(responseBody)
-      decode[A](responseBody) match {
-        case Left(err) => throw decodeError(responseBody).right.getOrElse(
-          WebServiceHelperError[A](response.code(), responseBody)
-        )
-        case Right(value) => value
+      decodedResponse <- {
```
Comment (tomrf1) [inline support-services/src/main/scala/com/gu/helpers/WebServiceHelper.scala]: can this block be moved out into a function? Nested for/yields can be hard to follow, especially in github

Comment (johnduffell) [reply support-services/src/main/scala/com/gu/helpers/WebServiceHelper.scala]: good idea hadn't spotted how long that had got

File: support-services/src/main/scala/com/gu/helpers/WebServiceHelper.scala [hunk 9]
```diff
@@ -51,19 +65,32 @@ trait WebServiceHelper[Error <: Throwable] {
...
+        val code = response.code().toString
+        val failableDecodedResponse = for {
+          body <- Option(response.body).toRight(WebServiceHelperError(code, "", "no body")).toTry
+          contentType <- Option(body.contentType()).toRight(WebServiceHelperError(code, "", "no content type")).toTry
+          bodyString <- Try(body.string())
+          _ = SafeLogger.info(s"response $code body: ${bodyString.length} bytes")
+          _ <- if (contentType.`type`() == "application" && contentType.subtype() == "json") Success(())
+          else Failure(WebServiceHelperError(code, bodyString, s"wrong content type"))
+          result <- code.splitAt(1) match {
```
Comment (tomrf1) [inline support-services/src/main/scala/com/gu/helpers/WebServiceHelper.scala]: I think this match block could also be a separate function, there's a lot going on in this `for` block

File: support-workers/src/main/scala/com/gu/support/workers/exceptions/RetryImplicits.scala (hunk 7 above)
Comment (johnduffell) [reply support-workers/src/main/scala/com/gu/support/workers/exceptions/RetryImplicits.scala]: TODO- I need to check that these still work (I think they don't in this PR)
https://github.com/guardian/support-frontend/blob/ef690cebb233c7ed593840937cb3b78d2d9cc800/support-models/src/main/scala/com/gu/support/zuora/api/response/Responses.scala#L48

File: support-services/src/main/scala/com/gu/helpers/WebServiceHelper.scala (hunk 6 above)
Comment (jfsoul) [reply support-services/src/main/scala/com/gu/helpers/WebServiceHelper.scala]: Maybe it would be easier to read quickly as an `if...else` with `code >= 200 && code < 300`? Separating out as Tom suggested is probably more meaningful though.

Comment (johnduffell) [reply support-services/src/main/scala/com/gu/helpers/WebServiceHelper.scala]:
```scala
      result <-
        if (code >= 200 && code < 300) 
          decode[A](bodyString).left.map { err =>
            WebServiceHelperError[A](code, bodyString, s"failed to parse response: $err")
          }.toTry
        else if (code >= 400 && code < 500) 
  Failure((decodeError(bodyString).right.toOption).getOrElse(
            WebServiceClientError(code, bodyString))
          )
        else Failure(WebServiceHelperError[A](code, bodyString, "unrecognised response code"))
      }
```

Comment (johnduffell) [reply support-services/src/main/scala/com/gu/helpers/WebServiceHelper.scala]: see what you mean about the WTF factor, I have an idea to make it read better

Comment (johnduffell) [reply support-services/src/main/scala/com/gu/helpers/WebServiceHelper.scala]:
```scala
      result <- code(0) + "xx" match {
        case "2xx" => decode[A](bodyString).left.map { err =>
          WebServiceHelperError[A](code, bodyString, s"failed to parse response: $err")
        }.toTry
        case "4xx" => Failure((decodeError(bodyString).right.toOption).getOrElse(
          WebServiceClientError(code, bodyString))
        )
        case _ => Failure(WebServiceHelperError[A](code, bodyString, "unrecognised response code"))
      }
```

File: support-services/src/main/scala/com/gu/helpers/WebServiceHelper.scala (hunk 5 above)
Comment (johnduffell) [reply support-services/src/main/scala/com/gu/helpers/WebServiceHelper.scala]: sorry, body is nullable, but string might throw an IOException.  In any case we probably just want to blow up and retry with a RetryUnlimited

File: support-workers/src/main/scala/com/gu/support/workers/exceptions/RetryImplicits.scala (hunk 7 above)
Comment (johnduffell) [reply support-workers/src/main/scala/com/gu/support/workers/exceptions/RetryImplicits.scala]: I have added the tests to ZuoraErrorsSpec.scala  below and they are now passing

**Comment by johnduffell** [comment]
I think this is ready, @jfsoul @tomrf1 if you wanted to take a look please 👀

**Comment by johnduffell** [comment]
tested in CODE, digi pack and contributions both worked

**Comment by johnduffell** [comment]
thanks Jon I'll merge that tomorrow morning
